crates/wasi-common/src/sys/windows/hostcalls_impl/misc.rs

```diff
@@ -77,12 +77,153 @@ pub(crate) fn clock_time_get(clock_id: wasi::__wasi_clockid_t) -> Result<wasi::_
     duration.as_nanos().try_into().map_err(Into::into)
 }
 
+fn stdin_nonempty() -> bool {
+    use std::io::Read;
+    std::io::stdin().bytes().peekable().peek().is_some()
+}
+
 pub(crate) fn poll_oneoff(
     timeout: Option<ClockEventData>,
     fd_events: Vec<FdEventData>,
     events: &mut Vec<wasi::__wasi_event_t>,
-) -> Result<Vec<wasi::__wasi_event_t>> {
-    unimplemented!("poll_oneoff")
+) -> Result<()> {
+    use crate::fdentry::Descriptor;
+    if fd_events.is_empty() && timeout.is_none() {
+        return Ok(());
+    }
+
+    // Currently WASI file support is only (a) regular files (b) directories (c) symlinks on Windows,
+    // which are always ready to write on Unix.
+    //
+    // We need to consider stdin/stdout/stderr separately. We treat stdout/stderr as always ready to write
+    // and only poll the stdin.
+    let mut stdin_events = vec![];
+    let mut immediate_events = vec![];
+
+    for event in fd_events {
+        match event.descriptor {
+            Descriptor::Stdin if event.r#type == wasi::__WASI_EVENTTYPE_FD_READ => {
+                stdin_events.push(event)
+            }
+            _ => immediate_events.push(event),
+        }
+    }
+
+    // we have at least one immediate event, so we don't need to care about stdin
+    if immediate_events.len() > 0 {
+        for event in immediate_events {
+            let size = match event.descriptor {
+                Descriptor::OsFile(os_file) if event.r#type == wasi::__WASI_EVENTTYPE_FD_READ => {
+                    os_file
+                        .metadata()
+                        .expect("FIXME return a proper error")
+                        .len()
+                }
+                Descriptor::Stdin => panic!("Descriptor::Stdin should have been filtered out"),
+                // On Unix, ioctl(FIONREAD) will return 0 for stdout/stderr. Emulate the same behavior on Windows.
+                //
+                // Besides, the spec is unclear what nbytes should actually be for __WASI_EVENTTYPE_FD_WRITE and
+                // the implementation on Unix just returns 0 here, so it's probably fine to do the same on Windows for now.
+                // cf. https://github.com/WebAssembly/WASI/issues/148
+                _ => 0,
+            };
+
+            events.push(wasi::__wasi_event_t {
+                userdata: event.userdata,
+                r#type: event.r#type,
+                error: wasi::__WASI_ESUCCESS,
+                u: wasi::__wasi_event_u {
+                    fd_readwrite: wasi::__wasi_event_fd_readwrite_t {
+                        nbytes: size,
+                        flags: 0,
+                    },
+                },
+            })
+        }
+    } else {
+        assert_ne!(stdin_events.len(), 0, "stdin_events should not be empty");
+
+        // We are busy-polling the stdin with delay, unfortunately.
```

Comment:
I think our best shot of implementing a `poll` for read of `stdin` would be a simple `WaitForSingleObject` on its handle; it should get signaled when there is data to read.

Reply:
The previous version of this change, 6245e432, actually mentioned `WaitForSingleObject. We can't do it unless we change the implementaion of `fd_read` and `fd_write`, but please give me a while to properly describe it.

Reply:
Are you sure that the wait functions would signal that stdout is ready for writing? They only seem to be mentioned for [low level input functions](https://docs.microsoft.com/en-us/windows/console/low-level-console-input-functions)

Reply:
For stdout/stderr, I'm not sure, but I would be very surprised if it didn't work.

For stdin, I'm confident that you can wait on the handle to signal ready to read.

Reply:
@peterhuene could you please take a look at the issues I've had with using `WaitForSingleObject`? In particular, can you reproduce them? They were mentioned in the OP. 

With them unsolved, `WaitForSingleObject` is useless anyway. Especially the timeout issue is a dealbreaker.

Reply:
I wrote a test case for this but was seeing unreliable behavior after the handle enters a signaled state, confirming the problems you were seeing.

I'd like to do a little more research here before we resign ourselves to spinning up a thread to implement this, which I'm not a fan of.

Reply:
Are you aware of the fact that I uploaded my test case, wait.cpp, in the OP?

Neither am I :)

Reply:
Yep, saw your test case (thanks!) and I was playing around with something similar. with and without the mode change.

Almost ready to give up on this pursuit :)

Reply:
Alright, I've played with this enough.

I think you've summed it up just fine in in the comments here regarding the pitfalls of `WaitForSingleObject` here.  In fact, with the default mode of `ENABLE_LINE_INPUT` set for the console, I think even `ReadFile` will buffer internally, causing it to potentially consume more than the requested console events, thereby resetting the signaled state of the stdin handle despite the fact a subsequent `ReadFile` may still return data.

I don't see around a busy poll implementation for this on Windows.  The plus side is that this should also work with redirected stdin, whereby any solution involving the Console API functions would fail.

Reply:
Resolving this discussion and I'll finish my review of the draft tomorrow morning PST (I need sleep to dream of a world where Windows works in a sane fashion).